tests/components/permobil/test_config_flow.py

```diff
@@ -0,0 +1,285 @@
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_CODE: MOCK_CODE},
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "email_code"
+    assert result["errors"]["base"] == "invalid_code"
+
+
+async def test_config_flow_incorrect_region(hass: HomeAssistant) -> None:
+    """Test the config flow from start to until the request for email code and have the API return error."""
+    mock_api: AsyncMock = AsyncMock()
+    mock_api.request_region_names.return_value = MOCK_REGIONS
+    mock_api.request_application_code.side_effect = MyPermobilAPIException
+    mock_api.request_application_token.return_value = None
+    # init flow
+    with patch(
+        "homeassistant.components.permobil.config_flow.MyPermobil",
+        return_value=mock_api,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            config_flow.DOMAIN,
+            context={"source": "user"},
+            data={CONF_EMAIL: MOCK_EMAIL},
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "region"
+    assert result["errors"] == {}
+    assert config_flow.PermobilConfigFlow.data[CONF_EMAIL] == MOCK_EMAIL
+
+    # select region step
+    # here the request_application_code raises a MyPermobilAPIException
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_REGION: MOCK_REGION_NAME},
+    )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "region"
+    assert result["errors"]["base"] == "code_request_error"
+
+
+async def test_config_flow_region_request_error(hass: HomeAssistant) -> None:
+    """Test the config flow from start to until the request for regions and have the API return error."""
+    mock_api: AsyncMock = AsyncMock()
+    mock_api.request_region_names.side_effect = MyPermobilAPIException
+    mock_api.request_application_code.return_value = None
+    mock_api.request_application_token.return_value = None
+    # init flow
+    # here the request_region_names raises a MyPermobilAPIException
+    with patch(
+        "homeassistant.components.permobil.config_flow.MyPermobil",
+        return_value=mock_api,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            config_flow.DOMAIN,
+            context={"source": "user"},
+            data={CONF_EMAIL: MOCK_EMAIL},
+        )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "region"
+    assert result["errors"]["base"] == "region_fetch_error"
+
+
+async def test_config_flow_invalid_email(hass: HomeAssistant) -> None:
+    """Test the config flow from start to until the request for regions and have the API return error."""
+
+    def set_email(_):
+        raise MyPermobilClientException
+
+    mock_api: AsyncMock = AsyncMock()
+    mock_api.set_email = set_email
```

Comment:
Why aren't you using `side_effect` here?

Reply:
For some reason doing `mock_api.set_email.side_effect = MyPermobilClientException` does not have the desired effect of of the error being raised, instead the execution just continued for some reason. I am still rather new to making tests so i was unsure how to solve it and this is what i came up with.

Reply:
Please see https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect